Bug Report: vtgate healthcheck: stale update from a replaced tablet's canceled `checkConn` can permanently remove a healthy, streaming replica from the healthy list

## Overview of the Issue

A race between `ReplaceTablet` and the replaced tablet's canceled `checkConn` goroutine can leave a vtgate permanently routing **zero queries** to a replica that is healthy, serving, and streaming health updates over a live connection to that same vtgate.

vtgate routes non-primary queries from `hc.healthy` only (`GetHealthyTabletStats`). That map is rebuilt exclusively by `recomputeHealthy`, which runs only on non-trivial updates and in `deleteTablet`; there is no periodic reconciliation. `hc.healthData` — which backs `/debug/healthcheck`, `SHOW VITESS_TABLETS`, and the `HealthcheckConnections` stats — is refreshed on every update, including trivial ones. The two diverge permanently if the last non-trivial write for a target key is wrong.

That wrong write can happen as follows. When a tablet comes back under the same alias with a different address (e.g. a Kubernetes pod rescheduled with a new IP), the topology watcher sees the changed host/port map on the next refresh (`--tablet-refresh-interval`, default 1m, with the default `--tablet-refresh-known-tablets=true`) and calls `ReplaceTablet` (topology_watcher.go:248):

1. `ReplaceTablet` = `RemoveTablet` + `AddTablet`, two separate critical sections. `deleteTablet` cancels the old `tabletHealthCheck`; `AddTablet` registers a new one under the same alias and spawns its `checkConn`.
2. The old `checkConn` goroutine pops out of `stream()` with an error and runs its error path, which calls `hc.updateHealth(thc.SimpleCopy(), thc.Target, false, false)` (tablet_health_check.go:305) without re-checking its own context. Between its stream returning and acquiring `hc.mu` it does real work (gRPC conn close, logging, counters), so its arrival at the lock can be delayed arbitrarily.
3. Meanwhile the new connection receives `Serving: true` (vttablet sends its current state immediately on subscribe). Non-trivial → `recomputeHealthy` → tablet healthy.
4. The old goroutine's final `updateHealth(Serving: false)` now lands. The deleted-tablet guard (healthcheck.go:535) only checks that the alias exists in `healthByAlias` — it cannot tell the update came from the dead `tabletHealthCheck` rather than the currently registered one. The stale write overwrites `healthData`, and `recomputeHealthy` drops the tablet from `healthy`.
5. Every subsequent response on the new connection is a `trivialUpdate`, because triviality is computed from the new `tabletHealthCheck`'s own struct fields (`thc.Serving` is still `true`; the stale write poisoned the maps, not the struct). Trivial updates refresh `healthData` — so the tablet looks healthy everywhere — but never run `recomputeHealthy`. The tablet stays out of the healthy (routing) list until an unrelated non-trivial event hits the same `keyspace/shard/type` key (another tablet's serving flap, lag crossing `--discovery_low_replication_lag`, a tablet added or removed) or the vtgate restarts; the eventual silent self-heal makes the condition hard to diagnose.

The timeout path (tablet_health_check.go:316) issues the same unguarded final update and has the same hazard. The per-event window is milliseconds wide, but it is raced on every tablet address change × every vtgate watching that cell.

## Reproduction Steps

Deterministic regression test:

https://github.com/mhamza15/vitess/commit/04a993ca2827a0a2564468cc69bd90375d06783d

```sh
go test ./go/vt/discovery/ -run TestStaleUpdateFromCanceledCheckConn -count=1
```

```
--- FAIL: TestStaleUpdateFromCanceledCheckConn (0.00s)
    Error:    "[]" should have 1 item(s), but has 0
    Messages: tablet is serving and streaming health but missing from the healthy
              list: stale update from the canceled checkConn poisoned the healthy list
```

How the interleaving is pinned down, and the caveats that come with that:

- The stale write is issued by the real `checkConn` error path; nothing is injected.
- The transport is the package's standard `fakeConn`, with one behavioral change for the old tablet's stream: on cancellation it returns an error (as a canceled gRPC stream does) instead of the fake's usual `nil`, and it parks until the test releases it. That gate is the only scheduling artifact: it selects the ordering in which the dying goroutine's final `updateHealth` lands after the replacement connection's first one. Nothing synchronizes those two writes in production — both contend on `hc.mu` with no ordering between them — so the test forces a schedule the runtime is otherwise free to produce; what it cannot show is how often production scheduling produces it.

<details>
<summary>Flaky end-to-end test</summary>

https://github.com/mhamza15/vitess/commit/77d4e6f48fa30149e4aa21f7e2d0a887dd901d5f

```sh
go test -count=1 -timeout 20m -v -run ^TestReplicaRoutableAfterReplaceTablet$ vitess.io/vitess/go/test/endtoend/tabletgateway/healthcheckrace
```

Output of the reproducing run, trimmed (`[...]` marks elisions):

```
=== RUN   TestReplicaRoutableAfterReplaceTablet
    stale_route_test.go:166: iteration 6: @replica queries failed continuously for 15s (149 failures)
    stale_route_test.go:167: first error: target: ks.0.replica: no healthy tablet available for 'keyspace:"ks" shard:"0" tablet_type:REPLICA' (errno 1105) (sqlstate HY000) during query: select id from t1
    stale_route_test.go:168: last error: target: ks.0.replica: no healthy tablet available for 'keyspace:"ks" shard:"0" tablet_type:REPLICA' (errno 1105) (sqlstate HY000) during query: select id from t1
    stale_route_test.go:169: vtgate routing answer for the replica target: target: ks.0.replica: no healthy tablet available for 'keyspace:"ks" shard:"0" tablet_type:REPLICA' (errno 1105) (sqlstate HY000) during query: select id from t1
    stale_route_test.go:170: @primary queries still work: true (error if not: <nil>)
    stale_route_test.go:171: healthcheck cache still counts the replica as serving: true
    stale_route_test.go:172: vtgate /debug/gateway (healthcheck cache view):
        [...]
          "TabletsStats": [
           {
            "Tablet": {
             "alias": {
              "cell": "zone1",
              "uid": 5146
             },
             [...]
            },
            "Target": {
             "keyspace": "ks",
             "shard": "0",
             "tablet_type": 2
            },
            "Serving": true,
            "PrimaryTermStartTime": 0,
            "Stats": {},
            "LastError": null
           }
          ]
        [...]
    stale_route_test.go:173: vtgate healthcheck log fingerprint (last lines):
        [...]
        2026-06-10 20:45:29.073 EDT INF discovery/healthcheck.go:474 Removing tablet from healthcheck: alias:{cell:"zone1" uid:5146} [...]
        2026-06-10 20:45:29.073 EDT INF discovery/healthcheck.go:420 Adding tablet to healthcheck: alias:{cell:"zone1" uid:5146} [...]
        2026-06-10 20:45:29.073 EDT WRN discovery/tablet_health_check.go:336 tablet alias:{cell:"zone1" uid:5146} [...] healthcheck stream error: Code: CANCELED
        2026-06-10 20:45:29.073 EDT INF discovery/tablet_health_check.go:112 HealthCheckUpdate(Serving State): tablet: zone1-5146 (localhost) serving true => false for ks/0 (REPLICA) reason: vttablet: rpc error: code = Canceled desc = context canceled
        2026-06-10 20:45:29.074 EDT INF discovery/tablet_health_check.go:112 HealthCheckUpdate(Serving State): tablet: zone1-5146 (localhost) serving false => true for ks/0 (REPLICA) reason: healthCheck update
    stale_route_test.go:182: 
        	Error:      	stale health update race reproduced
        	Test:       	TestReplicaRoutableAfterReplaceTablet
        	Messages:   	replica cell:"zone1" uid:5146 is serving and streaming health (per the healthcheck cache) but stayed out of the healthy list for 15s after ReplaceTablet; vtgate answers: target: ks.0.replica: no healthy tablet available for 'keyspace:"ks" shard:"0" tablet_type:REPLICA' (errno 1105) (sqlstate HY000) during query: select id from t1
--- FAIL: TestReplicaRoutableAfterReplaceTablet (17.66s)
FAIL
```

These are the last health transitions vtgate ever logged for the replica: the replace pair, the canceled stream's `serving true => false (context canceled)`, the replacement connection's `serving false => true` — then nothing for the 15 seconds during which routing kept answering "no healthy tablet". The transition lines record each goroutine's own state change before it reaches the healthcheck maps; the map write order, which is what the race decides, is what the sustained routing answer demonstrates.

</details>

## Binary Version

main @ 3140c1846cd120db34f33b2083139b4ab638aed2

## Operating System and Environment details

All.

## Log Fragments

N/A

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -199,6 +199,11 @@ type (
 	tabletAliasString       string
 )
 
+// aliasKey returns the map key used to index a tablet by its alias.
+func aliasKey(alias *topodata.TabletAlias) tabletAliasString {
+	return tabletAliasString(topoproto.TabletAliasString(alias))
+}
+
 // HealthCheck declares what the TabletGateway needs from the HealthCheck
 type HealthCheck interface {
 	// AddTablet adds the tablet.
@@ -475,7 +480,7 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	tabletAlias := tabletAliasString(topoproto.TabletAliasString(tablet.Alias))
+	tabletAlias := aliasKey(tablet.Alias)
 	defer func() {
 		// We want to be sure the tablet is gone from the secondary
 		// maps even if it's already gone from the authoritative map.
@@ -500,7 +505,7 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 					// clear the healthy list for the primary.
 					//
 					// See the logic in `updateHealth` for more details.
-					alias := tabletAliasString(topoproto.TabletAliasString(healthy[0].Tablet.Alias))
+					alias := aliasKey(healthy[0].Tablet.Alias)
 					if alias == tabletAlias {
 						hc.healthy[key] = []*TabletHealth{}
 					}
@@ -528,7 +533,7 @@ func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *quer
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	tabletAlias := tabletAliasString(topoproto.TabletAliasString(thc.Tablet.Alias))
+	tabletAlias := aliasKey(thc.Tablet.Alias)
 	// let's be sure that this tablet hasn't been deleted from the authoritative map
 	// so that we're not racing to update it and in effect re-adding a copy of the
 	// tablet record that was deleted. Also verify that the update was issued by the
@@ -605,7 +610,7 @@ func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *quer
 	case isPrimary && !up:
 		if healthy, ok := hc.healthy[targetKey]; ok && len(healthy) > 0 {
 			// isPrimary is true here therefore we should only have 1 tablet in healthy
-			alias := tabletAliasString(topoproto.TabletAliasString(healthy[0].Tablet.Alias))
+			alias := aliasKey(healthy[0].Tablet.Alias)
 			// Clear healthy list for primary if the existing tablet is down
 			if alias == tabletAlias {
 				hc.healthy[targetKey] = []*TabletHealth{}
@@ -883,7 +888,7 @@ func (hc *HealthCheckImpl) GetTabletHealthByAlias(alias *topodata.TabletAlias) (
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	if hd, ok := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(alias))]; ok {
+	if hd, ok := hc.healthByAlias[aliasKey(alias)]; ok {
 		return hd.SimpleCopy(), nil
 	}
 	return nil, fmt.Errorf("could not find tablet: %s", alias.String())
@@ -896,7 +901,7 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 	defer hc.mu.Unlock()
 
 	if hd, ok := hc.healthData[kst]; ok {
-		if th, ok := hd[tabletAliasString(topoproto.TabletAliasString(alias))]; ok {
+		if th, ok := hd[aliasKey(alias)]; ok {
 			return th, nil
 		}
 	}
@@ -908,7 +913,7 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 func (hc *HealthCheckImpl) registeredHealthCheck(alias *topodata.TabletAlias) *tabletHealthCheck {
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
-	return hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(alias))]
+	return hc.healthByAlias[aliasKey(alias)]
 }
 
 // TabletConnection returns the Connection to a given tablet.

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -524,23 +524,12 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 }
 
 func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *query.Target, trivialUpdate bool, up bool) {
-	// hc.healthByAlias is authoritative, it should be updated
 	tabletAlias := tabletAliasString(topoproto.TabletAliasString(thc.Tablet.Alias))
-	// let's be sure that this tablet hasn't been deleted from the authoritative map
-	// so that we're not racing to update it and in effect re-adding a copy of the
-	// tablet record that was deleted. Also verify that the update was issued by the
-	// tabletHealthCheck currently registered for this alias. After a ReplaceTablet,
-	// the canceled checkConn goroutine of the replaced tablet can still deliver one
-	// final update, which must not overwrite the state reported by its replacement.
-	if !hc.isRegisteredHealthCheck(tabletAlias, thc) {
-		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", thc.Tablet)
-		return
-	}
-
 	th := thc.SimpleCopy()
 
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
+	// Verify that this update came from the tabletHealthCheck currently registered for this alias.
 	if !hc.isRegisteredHealthCheckLocked(tabletAlias, thc) {
 		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", thc.Tablet)
 		return
@@ -906,13 +895,6 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 		}
 	}
 	return nil, fmt.Errorf("could not find tablet: %s", alias.String())
-}
-
-// isRegisteredHealthCheck reports whether thc is the current healthcheck for tabletAlias.
-func (hc *HealthCheckImpl) isRegisteredHealthCheck(tabletAlias tabletAliasString, thc *tabletHealthCheck) bool {
-	hc.mu.Lock()
-	defer hc.mu.Unlock()
-	return hc.isRegisteredHealthCheckLocked(tabletAlias, thc)
 }
 
 // isRegisteredHealthCheckLocked reports whether thc is the current healthcheck for tabletAlias.

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -199,11 +199,6 @@ type (
 	tabletAliasString       string
 )
 
-// aliasKey returns the map key used to index a tablet by its alias.
-func aliasKey(alias *topodata.TabletAlias) tabletAliasString {
-	return tabletAliasString(topoproto.TabletAliasString(alias))
-}
-
 // HealthCheck declares what the TabletGateway needs from the HealthCheck
 type HealthCheck interface {
 	// AddTablet adds the tablet.
@@ -480,7 +475,7 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	tabletAlias := aliasKey(tablet.Alias)
+	tabletAlias := tabletAliasString(topoproto.TabletAliasString(tablet.Alias))
 	defer func() {
 		// We want to be sure the tablet is gone from the secondary
 		// maps even if it's already gone from the authoritative map.
@@ -505,7 +500,7 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 					// clear the healthy list for the primary.
 					//
 					// See the logic in `updateHealth` for more details.
-					alias := aliasKey(healthy[0].Tablet.Alias)
+					alias := tabletAliasString(topoproto.TabletAliasString(healthy[0].Tablet.Alias))
 					if alias == tabletAlias {
 						hc.healthy[key] = []*TabletHealth{}
 					}
@@ -533,7 +528,7 @@ func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *quer
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	tabletAlias := aliasKey(thc.Tablet.Alias)
+	tabletAlias := tabletAliasString(topoproto.TabletAliasString(thc.Tablet.Alias))
 	// let's be sure that this tablet hasn't been deleted from the authoritative map
 	// so that we're not racing to update it and in effect re-adding a copy of the
 	// tablet record that was deleted. Also verify that the update was issued by the
@@ -610,7 +605,7 @@ func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *quer
 	case isPrimary && !up:
 		if healthy, ok := hc.healthy[targetKey]; ok && len(healthy) > 0 {
 			// isPrimary is true here therefore we should only have 1 tablet in healthy
-			alias := aliasKey(healthy[0].Tablet.Alias)
+			alias := tabletAliasString(topoproto.TabletAliasString(healthy[0].Tablet.Alias))
 			// Clear healthy list for primary if the existing tablet is down
 			if alias == tabletAlias {
 				hc.healthy[targetKey] = []*TabletHealth{}
@@ -888,7 +883,7 @@ func (hc *HealthCheckImpl) GetTabletHealthByAlias(alias *topodata.TabletAlias) (
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	if hd, ok := hc.healthByAlias[aliasKey(alias)]; ok {
+	if hd, ok := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(alias))]; ok {
 		return hd.SimpleCopy(), nil
 	}
 	return nil, fmt.Errorf("could not find tablet: %s", alias.String())
@@ -901,7 +896,7 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 	defer hc.mu.Unlock()
 
 	if hd, ok := hc.healthData[kst]; ok {
-		if th, ok := hd[aliasKey(alias)]; ok {
+		if th, ok := hd[tabletAliasString(topoproto.TabletAliasString(alias))]; ok {
 			return th, nil
 		}
 	}
@@ -913,7 +908,7 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 func (hc *HealthCheckImpl) registeredHealthCheck(alias *topodata.TabletAlias) *tabletHealthCheck {
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
-	return hc.healthByAlias[aliasKey(alias)]
+	return hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(alias))]
 }
 
 // TabletConnection returns the Connection to a given tablet.

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -903,11 +903,17 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 	return nil, fmt.Errorf("could not find tablet: %s", alias.String())
 }
 
+// registeredHealthCheck returns the tabletHealthCheck currently registered
+// for the given alias, or nil if there is none.
+func (hc *HealthCheckImpl) registeredHealthCheck(alias *topodata.TabletAlias) *tabletHealthCheck {
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	return hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(alias))]
+}
+
 // TabletConnection returns the Connection to a given tablet.
 func (hc *HealthCheckImpl) TabletConnection(ctx context.Context, alias *topodata.TabletAlias, target *query.Target) (queryservice.QueryService, error) {
-	hc.mu.Lock()
-	thc := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(alias))]
-	hc.mu.Unlock()
+	thc := hc.registeredHealthCheck(alias)
 	if thc == nil || thc.Conn == nil {
 		return nil, vterrors.Errorf(vtrpc.Code_NOT_FOUND, "tablet: %v is either down or nonexistent", alias)
 	}

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -525,9 +525,6 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 
 func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *query.Target, trivialUpdate bool, up bool) {
 	// hc.healthByAlias is authoritative, it should be updated
-	hc.mu.Lock()
-	defer hc.mu.Unlock()
-
 	tabletAlias := tabletAliasString(topoproto.TabletAliasString(thc.Tablet.Alias))
 	// let's be sure that this tablet hasn't been deleted from the authoritative map
 	// so that we're not racing to update it and in effect re-adding a copy of the
@@ -535,12 +532,20 @@ func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *quer
 	// tabletHealthCheck currently registered for this alias. After a ReplaceTablet,
 	// the canceled checkConn goroutine of the replaced tablet can still deliver one
 	// final update, which must not overwrite the state reported by its replacement.
-	if registered, ok := hc.healthByAlias[tabletAlias]; !ok || registered != thc {
+	if !hc.isRegisteredHealthCheck(tabletAlias, thc) {
 		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", thc.Tablet)
 		return
 	}
 
 	th := thc.SimpleCopy()
+
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	if !hc.isRegisteredHealthCheckLocked(tabletAlias, thc) {
+		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", thc.Tablet)
+		return
+	}
+
 	targetKey := KeyFromTarget(th.Target)
 	targetChanged := prevTarget.TabletType != th.Target.TabletType || prevTarget.Keyspace != th.Target.Keyspace || prevTarget.Shard != th.Target.Shard
 	if targetChanged {
@@ -901,6 +906,20 @@ func (hc *HealthCheckImpl) GetTabletHealth(kst KeyspaceShardTabletType, alias *t
 		}
 	}
 	return nil, fmt.Errorf("could not find tablet: %s", alias.String())
+}
+
+// isRegisteredHealthCheck reports whether thc is the current healthcheck for tabletAlias.
+func (hc *HealthCheckImpl) isRegisteredHealthCheck(tabletAlias tabletAliasString, thc *tabletHealthCheck) bool {
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	return hc.isRegisteredHealthCheckLocked(tabletAlias, thc)
+}
+
+// isRegisteredHealthCheckLocked reports whether thc is the current healthcheck for tabletAlias.
+// hc.mu must be locked before calling it.
+func (hc *HealthCheckImpl) isRegisteredHealthCheckLocked(tabletAlias tabletAliasString, thc *tabletHealthCheck) bool {
+	registered, ok := hc.healthByAlias[tabletAlias]
+	return ok && registered == thc
 }
 
 // registeredHealthCheck returns the tabletHealthCheck currently registered

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -523,12 +523,12 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 	delete(hc.healthByAlias, tabletAlias)
 }
 
-func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, th *TabletHealth, prevTarget *query.Target, trivialUpdate bool, up bool) {
+func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, prevTarget *query.Target, trivialUpdate bool, up bool) {
 	// hc.healthByAlias is authoritative, it should be updated
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 
-	tabletAlias := tabletAliasString(topoproto.TabletAliasString(th.Tablet.Alias))
+	tabletAlias := tabletAliasString(topoproto.TabletAliasString(thc.Tablet.Alias))
 	// let's be sure that this tablet hasn't been deleted from the authoritative map
 	// so that we're not racing to update it and in effect re-adding a copy of the
 	// tablet record that was deleted. Also verify that the update was issued by the
@@ -536,10 +536,11 @@ func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, th *TabletHealth
 	// the canceled checkConn goroutine of the replaced tablet can still deliver one
 	// final update, which must not overwrite the state reported by its replacement.
 	if registered, ok := hc.healthByAlias[tabletAlias]; !ok || registered != thc {
-		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", th.Tablet)
+		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", thc.Tablet)
 		return
 	}
 
+	th := thc.SimpleCopy()
 	targetKey := KeyFromTarget(th.Target)
 	targetChanged := prevTarget.TabletType != th.Target.TabletType || prevTarget.Keyspace != th.Target.Keyspace || prevTarget.Shard != th.Target.Shard
 	if targetChanged {

--- a/go/vt/discovery/healthcheck.go
+++ b/go/vt/discovery/healthcheck.go
@@ -523,7 +523,7 @@ func (hc *HealthCheckImpl) deleteTablet(tablet *topodata.Tablet) {
 	delete(hc.healthByAlias, tabletAlias)
 }
 
-func (hc *HealthCheckImpl) updateHealth(th *TabletHealth, prevTarget *query.Target, trivialUpdate bool, up bool) {
+func (hc *HealthCheckImpl) updateHealth(thc *tabletHealthCheck, th *TabletHealth, prevTarget *query.Target, trivialUpdate bool, up bool) {
 	// hc.healthByAlias is authoritative, it should be updated
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
@@ -531,9 +531,12 @@ func (hc *HealthCheckImpl) updateHealth(th *TabletHealth, prevTarget *query.Targ
 	tabletAlias := tabletAliasString(topoproto.TabletAliasString(th.Tablet.Alias))
 	// let's be sure that this tablet hasn't been deleted from the authoritative map
 	// so that we're not racing to update it and in effect re-adding a copy of the
-	// tablet record that was deleted
-	if _, ok := hc.healthByAlias[tabletAlias]; !ok {
-		hc.logger().Infof("Tablet %v has been deleted, skipping health update", th.Tablet)
+	// tablet record that was deleted. Also verify that the update was issued by the
+	// tabletHealthCheck currently registered for this alias. After a ReplaceTablet,
+	// the canceled checkConn goroutine of the replaced tablet can still deliver one
+	// final update, which must not overwrite the state reported by its replacement.
+	if registered, ok := hc.healthByAlias[tabletAlias]; !ok || registered != thc {
+		hc.logger().Infof("Tablet %v has been deleted or replaced, skipping health update", th.Tablet)
 		return
 	}
 

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -27,6 +27,7 @@ import (
 	"sync"
 	"sync/atomic"
 	"testing"
+	"testing/synctest"
 	"time"
 
 	"github.com/google/safehtml/template"
@@ -49,7 +50,7 @@ import (
 )
 
 var (
-	connMap   map[string]*fakeConn
+	connMap   map[string]queryservice.QueryService
 	connMapMu sync.Mutex
 )
 
@@ -61,7 +62,7 @@ func testChecksum(t *testing.T, want, got int64) {
 func init() {
 	tabletconn.RegisterDialer("fake_gateway", tabletDialer)
 	tabletconntest.SetProtocol("go.vt.discovery.healthcheck_test", "fake_gateway")
-	connMap = make(map[string]*fakeConn)
+	connMap = make(map[string]queryservice.QueryService)
 	refreshInterval = time.Minute
 }
 
@@ -855,6 +856,136 @@ func TestRemoveTablet(t *testing.T) {
 	assert.Empty(t, a, "wrong result, expected empty list")
 }
 
+// TestStaleUpdateFromCanceledCheckConn is a regression test for a race
+// between ReplaceTablet and the replaced tablet's checkConn goroutine.
+//
+// When the topology watcher observes that a tablet's address changed (same
+// alias, new host or port map), it calls ReplaceTablet, which cancels the
+// old tabletHealthCheck and registers a new one under the same alias. The
+// old checkConn goroutine reacts to the cancellation from inside stream()
+// and issues one final updateHealth(Serving: false) from its error path
+// without re-checking its own context. Because updateHealth only verifies
+// that the alias still exists in healthByAlias — not that the update came
+// from the currently registered tabletHealthCheck — that stale update can be
+// applied after the replacement connection has already reported
+// Serving: true:
+//
+//  1. ReplaceTablet cancels the old checkConn and registers a new
+//     tabletHealthCheck for the same alias.
+//  2. The new connection receives Serving: true; the tablet is added to the
+//     healthy list.
+//  3. The old goroutine's final updateHealth(Serving: false) lands, removing
+//     the tablet from the healthy list and poisoning healthData.
+//  4. Every subsequent response on the new connection is a trivialUpdate
+//     (the new tabletHealthCheck's own Serving field is still true), so
+//     recomputeHealthy never runs again: healthData shows a healthy, serving
+//     tablet with a live stream while the healthy (routing) list permanently
+//     omits it. The tablet receives zero traffic until an unrelated
+//     non-trivial event in the same keyspace/shard/type triggers a
+//     recompute.
+//
+// Step 3's timing is scheduler-dependent (the dying goroutine can be delayed
+// between its stream returning and acquiring hc.mu), so the old connection's
+// stream is gated: the canceled checkConn goroutine stays parked inside
+// stream() until the test releases it, and synctest.Wait() guarantees each
+// step is fully applied before the next.
+func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ctx := t.Context()
+		ts := memorytopo.NewServer(ctx, "cell")
+		defer ts.Close()
+		hc := createTestHc(ctx, ts)
+		defer hc.Close()
+
+		target := &querypb.Target{Keyspace: "k", Shard: "s", TabletType: topodatapb.TabletType_REPLICA}
+
+		// The tablet before the restart, with its stream gated on cancellation.
+		oldTablet := createTestTablet(0, "cell", "a")
+		oldTablet.Type = topodatapb.TabletType_REPLICA
+		oldInput := make(chan *querypb.StreamHealthResponse)
+		oldConn := &staleStreamConn{
+			fakeConn: createFakeConn(oldTablet, oldInput),
+			release:  make(chan struct{}),
+		}
+		connMapMu.Lock()
+		connMap[TabletToMapKey(oldTablet)] = oldConn
+		connMapMu.Unlock()
+
+		// The same tablet after the restart: same alias, new ports.
+		newTablet := createTestTablet(0, "cell", "a")
+		newTablet.Type = topodatapb.TabletType_REPLICA
+		newTablet.PortMap["vt"] = 5
+		newTablet.PortMap["grpc"] = 6
+		newInput := make(chan *querypb.StreamHealthResponse)
+		createFakeConn(newTablet, newInput)
+
+		shr := &querypb.StreamHealthResponse{
+			TabletAlias:   oldTablet.Alias,
+			Target:        target,
+			Serving:       true,
+			RealtimeStats: &querypb.RealtimeStats{ReplicationLagSeconds: 1, CpuUsage: 0.2},
+		}
+
+		hc.AddTablet(oldTablet)
+		oldInput <- shr
+		synctest.Wait()
+		require.Len(t, hc.GetHealthyTabletStats(target), 1, "tablet should be healthy before the restart")
+
+		// The topology watcher notices the restarted tablet and replaces it.
+		// The old checkConn goroutine is canceled but stays parked inside its
+		// stream, before it runs its error path.
+		hc.ReplaceTablet(oldTablet, newTablet)
+
+		// The new connection reports Serving: true; the tablet becomes healthy.
+		newInput <- shr
+		synctest.Wait()
+		require.Len(t, hc.GetHealthyTabletStats(target), 1, "tablet should be healthy again after the replace")
+
+		// Release the canceled goroutine. Its stream now returns an error and
+		// the checkConn error path issues the final stale
+		// updateHealth(Serving: false) — after the replacement connection has
+		// already reported healthy.
+		close(oldConn.release)
+		synctest.Wait()
+
+		// The tablet keeps reporting good health on the new connection: a
+		// trivialUpdate, like the steady-state updates of a healthy replica.
+		newInput <- shr
+		synctest.Wait()
+
+		// The healthcheck cache (healthData) shows a serving tablet with a
+		// live stream, so the healthy (routing) list must include it as well.
+		healthy := hc.GetHealthyTabletStats(target)
+		require.Len(t, healthy, 1,
+			"tablet is serving and streaming health but missing from the healthy list: stale update from the canceled checkConn poisoned the healthy list")
+		assert.True(t, healthy[0].Serving)
+	})
+}
+
+// staleStreamConn wraps fakeConn so that a canceled stream parks until
+// release is closed and then fails like a canceled gRPC stream, modeling the
+// window between a checkConn goroutine's cancellation and its final
+// updateHealth.
+type staleStreamConn struct {
+	*fakeConn
+	release chan struct{}
+}
+
+// StreamHealth implements queryservice.QueryService.
+func (c *staleStreamConn) StreamHealth(ctx context.Context, callback func(shr *querypb.StreamHealthResponse) error) error {
+	for {
+		select {
+		case shr := <-c.hcChan:
+			if err := callback(shr); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			<-c.release
+			return ctx.Err()
+		}
+	}
+}
+
 // When an external primary failover is performed,
 // the demoted primary will advertise itself as a `PRIMARY`
 // tablet until it recognizes that it was demoted,
@@ -1268,13 +1399,18 @@ func TestLoadTabletsTrigger(t *testing.T) {
 	}
 	hc.AddTablet(tablet1)
 
+	hc.mu.Lock()
+	thc := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(tablet1.Alias))]
+	hc.mu.Unlock()
+	require.NotNil(t, thc)
+
 	numTriggers := 10
 	for range numTriggers {
 		// Since the previous target was a primary, and there are no other
 		// primary tablets for the given keyspace shard, we will see the healtcheck
 		// send on the loadTablets trigger. We just want to verify the information
 		// there is correct.
-		hc.updateHealth(th, prevTarget, false, false)
+		hc.updateHealth(thc, th, prevTarget, false, false)
 	}
 
 	ch := hc.GetLoadTabletsTrigger()

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -883,9 +883,9 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 	synctest.Test(t, func(t *testing.T) {
 		ctx := t.Context()
 		ts := memorytopo.NewServer(ctx, "cell")
-		defer ts.Close()
+		t.Cleanup(func() { ts.Close() })
 		hc := createTestHc(ctx, ts)
-		defer hc.Close()
+		t.Cleanup(func() { hc.Close() })
 
 		target := &querypb.Target{Keyspace: "k", Shard: "s", TabletType: topodatapb.TabletType_REPLICA}
 
@@ -895,12 +895,12 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 		oldInput := make(chan *querypb.StreamHealthResponse)
 		oldConn := createFakeConn(oldTablet, oldInput)
 		oldConn.releaseOnCancel = make(chan struct{})
-		// Always release the parked goroutine, even when an assertion fails
-		// before the inline release: the deferred hc.Close() waits for it, and
-		// a goroutine parked forever would turn the test failure into a
-		// synctest bubble-deadlock panic.
+		// Registered last so it runs before the hc.Close cleanup: hc.Close
+		// waits for this goroutine, and one parked forever would turn an early
+		// assertion failure into a synctest bubble-deadlock panic. OnceFunc
+		// guards against the double close from the deliberate inline release.
 		releaseOldConn := sync.OnceFunc(func() { close(oldConn.releaseOnCancel) })
-		defer releaseOldConn()
+		t.Cleanup(releaseOldConn)
 
 		// The same tablet after the restart: same alias, new ports.
 		newTablet := createTestTablet(0, "cell", "a")

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -910,6 +910,12 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 		connMapMu.Lock()
 		connMap[TabletToMapKey(oldTablet)] = oldConn
 		connMapMu.Unlock()
+		// Always release the parked goroutine, even when an assertion fails
+		// before the inline release: the deferred hc.Close() waits for it, and
+		// a goroutine parked forever would turn the test failure into a
+		// synctest bubble-deadlock panic.
+		releaseOldConn := sync.OnceFunc(func() { close(oldConn.release) })
+		defer releaseOldConn()
 
 		// The same tablet after the restart: same alias, new ports.
 		newTablet := createTestTablet(0, "cell", "a")
@@ -945,7 +951,7 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 		// the checkConn error path issues the final stale
 		// updateHealth(Serving: false) — after the replacement connection has
 		// already reported healthy.
-		close(oldConn.release)
+		releaseOldConn()
 		synctest.Wait()
 
 		// The tablet keeps reporting good health on the new connection: a

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -866,29 +866,19 @@ func TestRemoveTablet(t *testing.T) {
 // and issues one final updateHealth(Serving: false) from its error path
 // without re-checking its own context. Because updateHealth only verifies
 // that the alias still exists in healthByAlias — not that the update came
-// from the currently registered tabletHealthCheck — that stale update can be
-// applied after the replacement connection has already reported
-// Serving: true:
+// from the currently registered tabletHealthCheck — that stale update can
+// land after the replacement connection has already reported Serving: true,
+// removing the tablet from the healthy (routing) list and poisoning
+// healthData. Every subsequent response on the new connection is then a
+// trivialUpdate (the new tabletHealthCheck's own Serving field is still
+// true), so recomputeHealthy never runs again: the tablet receives zero
+// traffic until an unrelated non-trivial event in the same
+// keyspace/shard/type triggers a recompute.
 //
-//  1. ReplaceTablet cancels the old checkConn and registers a new
-//     tabletHealthCheck for the same alias.
-//  2. The new connection receives Serving: true; the tablet is added to the
-//     healthy list.
-//  3. The old goroutine's final updateHealth(Serving: false) lands, removing
-//     the tablet from the healthy list and poisoning healthData.
-//  4. Every subsequent response on the new connection is a trivialUpdate
-//     (the new tabletHealthCheck's own Serving field is still true), so
-//     recomputeHealthy never runs again: healthData shows a healthy, serving
-//     tablet with a live stream while the healthy (routing) list permanently
-//     omits it. The tablet receives zero traffic until an unrelated
-//     non-trivial event in the same keyspace/shard/type triggers a
-//     recompute.
-//
-// Step 3's timing is scheduler-dependent (the dying goroutine can be delayed
-// between its stream returning and acquiring hc.mu), so the old connection's
-// stream is gated: the canceled checkConn goroutine stays parked inside
-// stream() until the test releases it, and synctest.Wait() guarantees each
-// step is fully applied before the next.
+// The stale update's timing is scheduler-dependent (the dying goroutine can
+// be delayed between its stream returning and acquiring hc.mu), so the old
+// connection's stream is gated on cancellation, and synctest.Wait()
+// guarantees each step is fully applied before the next.
 func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 	synctest.Test(t, func(t *testing.T) {
 		ctx := t.Context()

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -1368,9 +1368,7 @@ func TestLoadTabletsTrigger(t *testing.T) {
 	}
 	hc.AddTablet(tablet1)
 
-	hc.mu.Lock()
-	thc := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(tablet1.Alias))]
-	hc.mu.Unlock()
+	thc := hc.registeredHealthCheck(tablet1.Alias)
 	require.NotNil(t, thc)
 
 	numTriggers := 10

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -939,11 +939,6 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 		releaseOldConn()
 		synctest.Wait()
 
-		// The tablet keeps reporting good health on the new connection: a
-		// trivialUpdate, like the steady-state updates of a healthy replica.
-		newInput <- shr
-		synctest.Wait()
-
 		// The healthcheck cache (healthData) shows a serving tablet with a
 		// live stream, so the healthy (routing) list must include it as well.
 		healthy := hc.GetHealthyTabletStats(target)

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -856,29 +856,9 @@ func TestRemoveTablet(t *testing.T) {
 	assert.Empty(t, a, "wrong result, expected empty list")
 }
 
-// TestStaleUpdateFromCanceledCheckConn is a regression test for a race
-// between ReplaceTablet and the replaced tablet's checkConn goroutine.
-//
-// When the topology watcher observes that a tablet's address changed (same
-// alias, new host or port map), it calls ReplaceTablet, which cancels the
-// old tabletHealthCheck and registers a new one under the same alias. The
-// old checkConn goroutine reacts to the cancellation from inside stream()
-// and issues one final updateHealth(Serving: false) from its error path
-// without re-checking its own context. Because updateHealth only verifies
-// that the alias still exists in healthByAlias — not that the update came
-// from the currently registered tabletHealthCheck — that stale update can
-// land after the replacement connection has already reported Serving: true,
-// removing the tablet from the healthy (routing) list and poisoning
-// healthData. Every subsequent response on the new connection is then a
-// trivialUpdate (the new tabletHealthCheck's own Serving field is still
-// true), so recomputeHealthy never runs again: the tablet receives zero
-// traffic until an unrelated non-trivial event in the same
-// keyspace/shard/type triggers a recompute.
-//
-// The stale update's timing is scheduler-dependent (the dying goroutine can
-// be delayed between its stream returning and acquiring hc.mu), so the old
-// connection's stream is gated on cancellation, and synctest.Wait()
-// guarantees each step is fully applied before the next.
+// TestStaleUpdateFromCanceledCheckConn checks that a replaced tablet's
+// canceled checkConn cannot remove its healthy replacement from the routing
+// list with a final, stale updateHealth(Serving: false).
 func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 	synctest.Test(t, func(t *testing.T) {
 		ctx := t.Context()
@@ -895,10 +875,6 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 		oldInput := make(chan *querypb.StreamHealthResponse)
 		oldConn := createFakeConn(oldTablet, oldInput)
 		oldConn.releaseOnCancel = make(chan struct{})
-		// Registered last so it runs before the hc.Close cleanup: hc.Close
-		// waits for this goroutine, and one parked forever would turn an early
-		// assertion failure into a synctest bubble-deadlock panic. OnceFunc
-		// guards against the double close from the deliberate inline release.
 		releaseOldConn := sync.OnceFunc(func() { close(oldConn.releaseOnCancel) })
 		t.Cleanup(releaseOldConn)
 

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -1359,16 +1359,8 @@ func TestLoadTabletsTrigger(t *testing.T) {
 		Shard:    shard,
 	}
 
-	// We want to run updateHealth with arguments that always
-	// make it trigger load Tablets.
-	th := &TabletHealth{
-		Tablet: tablet1,
-		Target: &querypb.Target{
-			Keyspace:   ks,
-			Shard:      shard,
-			TabletType: topodatapb.TabletType_REPLICA,
-		},
-	}
+	// We want to run updateHealth with a previous target that always
+	// makes it trigger load Tablets.
 	prevTarget := &querypb.Target{
 		Keyspace:   ks,
 		Shard:      shard,
@@ -1387,7 +1379,7 @@ func TestLoadTabletsTrigger(t *testing.T) {
 		// primary tablets for the given keyspace shard, we will see the healtcheck
 		// send on the loadTablets trigger. We just want to verify the information
 		// there is correct.
-		hc.updateHealth(thc, th, prevTarget, false, false)
+		hc.updateHealth(thc, prevTarget, false, false)
 	}
 
 	ch := hc.GetLoadTabletsTrigger()

--- a/go/vt/discovery/healthcheck_test.go
+++ b/go/vt/discovery/healthcheck_test.go
@@ -50,7 +50,7 @@ import (
 )
 
 var (
-	connMap   map[string]queryservice.QueryService
+	connMap   map[string]*fakeConn
 	connMapMu sync.Mutex
 )
 
@@ -62,7 +62,7 @@ func testChecksum(t *testing.T, want, got int64) {
 func init() {
 	tabletconn.RegisterDialer("fake_gateway", tabletDialer)
 	tabletconntest.SetProtocol("go.vt.discovery.healthcheck_test", "fake_gateway")
-	connMap = make(map[string]queryservice.QueryService)
+	connMap = make(map[string]*fakeConn)
 	refreshInterval = time.Minute
 }
 
@@ -903,18 +903,13 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 		oldTablet := createTestTablet(0, "cell", "a")
 		oldTablet.Type = topodatapb.TabletType_REPLICA
 		oldInput := make(chan *querypb.StreamHealthResponse)
-		oldConn := &staleStreamConn{
-			fakeConn: createFakeConn(oldTablet, oldInput),
-			release:  make(chan struct{}),
-		}
-		connMapMu.Lock()
-		connMap[TabletToMapKey(oldTablet)] = oldConn
-		connMapMu.Unlock()
+		oldConn := createFakeConn(oldTablet, oldInput)
+		oldConn.releaseOnCancel = make(chan struct{})
 		// Always release the parked goroutine, even when an assertion fails
 		// before the inline release: the deferred hc.Close() waits for it, and
 		// a goroutine parked forever would turn the test failure into a
 		// synctest bubble-deadlock panic.
-		releaseOldConn := sync.OnceFunc(func() { close(oldConn.release) })
+		releaseOldConn := sync.OnceFunc(func() { close(oldConn.releaseOnCancel) })
 		defer releaseOldConn()
 
 		// The same tablet after the restart: same alias, new ports.
@@ -966,30 +961,6 @@ func TestStaleUpdateFromCanceledCheckConn(t *testing.T) {
 			"tablet is serving and streaming health but missing from the healthy list: stale update from the canceled checkConn poisoned the healthy list")
 		assert.True(t, healthy[0].Serving)
 	})
-}
-
-// staleStreamConn wraps fakeConn so that a canceled stream parks until
-// release is closed and then fails like a canceled gRPC stream, modeling the
-// window between a checkConn goroutine's cancellation and its final
-// updateHealth.
-type staleStreamConn struct {
-	*fakeConn
-	release chan struct{}
-}
-
-// StreamHealth implements queryservice.QueryService.
-func (c *staleStreamConn) StreamHealth(ctx context.Context, callback func(shr *querypb.StreamHealthResponse) error) error {
-	for {
-		select {
-		case shr := <-c.hcChan:
-			if err := callback(shr); err != nil {
-				return err
-			}
-		case <-ctx.Done():
-			<-c.release
-			return ctx.Err()
-		}
-	}
 }
 
 // When an external primary failover is performed,
@@ -1805,6 +1776,10 @@ type fakeConn struct {
 	errCh chan error
 	// cbErrCh is a channel which receives errors returned from the supplied callback.
 	cbErrCh chan error
+	// releaseOnCancel, if non-nil, parks a canceled stream until the channel
+	// is closed and then makes it return ctx.Err() like a real gRPC stream,
+	// instead of returning nil right away.
+	releaseOnCancel chan struct{}
 
 	mu       sync.Mutex
 	canceled bool
@@ -1848,6 +1823,10 @@ func (fc *fakeConn) StreamHealth(ctx context.Context, callback func(shr *querypb
 			fc.mu.Lock()
 			fc.canceled = true
 			fc.mu.Unlock()
+			if fc.releaseOnCancel != nil {
+				<-fc.releaseOnCancel
+				return ctx.Err()
+			}
 			return nil
 		}
 	}

--- a/go/vt/discovery/tablet_health_check.go
+++ b/go/vt/discovery/tablet_health_check.go
@@ -201,7 +201,7 @@ func (thc *tabletHealthCheck) processResponse(hc *HealthCheckImpl, shr *query.St
 	thc.setServingState(serving, reason)
 
 	// notify downstream for primary change
-	hc.updateHealth(thc, thc.SimpleCopy(), prevTarget, trivialUpdate, thc.Serving)
+	hc.updateHealth(thc, prevTarget, trivialUpdate, thc.Serving)
 	return nil
 }
 
@@ -302,7 +302,7 @@ func (thc *tabletHealthCheck) checkConn(hc *HealthCheckImpl) {
 			}
 			// trivialUpdate = false because this is an error
 			// up = false because we did not get a healthy response
-			hc.updateHealth(thc, thc.SimpleCopy(), thc.Target, false, false)
+			hc.updateHealth(thc, thc.Target, false, false)
 		}
 		// If there was a timeout send an error. We do this after stream has returned.
 		// This will ensure that this update prevails over any previous message that
@@ -313,7 +313,7 @@ func (thc *tabletHealthCheck) checkConn(hc *HealthCheckImpl) {
 			hcErrorCounters.Add([]string{thc.Target.Keyspace, thc.Target.Shard, topoproto.TabletTypeLString(thc.Target.TabletType)}, 1)
 			// trivialUpdate = false because this is an error
 			// up = false because we did not get a healthy response within the timeout
-			hc.updateHealth(thc, thc.SimpleCopy(), thc.Target, false, false)
+			hc.updateHealth(thc, thc.Target, false, false)
 		}
 
 		// Streaming RPC failed e.g. because vttablet was restarted or took too long.

--- a/go/vt/discovery/tablet_health_check.go
+++ b/go/vt/discovery/tablet_health_check.go
@@ -201,7 +201,7 @@ func (thc *tabletHealthCheck) processResponse(hc *HealthCheckImpl, shr *query.St
 	thc.setServingState(serving, reason)
 
 	// notify downstream for primary change
-	hc.updateHealth(thc.SimpleCopy(), prevTarget, trivialUpdate, thc.Serving)
+	hc.updateHealth(thc, thc.SimpleCopy(), prevTarget, trivialUpdate, thc.Serving)
 	return nil
 }
 
@@ -302,7 +302,7 @@ func (thc *tabletHealthCheck) checkConn(hc *HealthCheckImpl) {
 			}
 			// trivialUpdate = false because this is an error
 			// up = false because we did not get a healthy response
-			hc.updateHealth(thc.SimpleCopy(), thc.Target, false, false)
+			hc.updateHealth(thc, thc.SimpleCopy(), thc.Target, false, false)
 		}
 		// If there was a timeout send an error. We do this after stream has returned.
 		// This will ensure that this update prevails over any previous message that
@@ -313,7 +313,7 @@ func (thc *tabletHealthCheck) checkConn(hc *HealthCheckImpl) {
 			hcErrorCounters.Add([]string{thc.Target.Keyspace, thc.Target.Shard, topoproto.TabletTypeLString(thc.Target.TabletType)}, 1)
 			// trivialUpdate = false because this is an error
 			// up = false because we did not get a healthy response within the timeout
-			hc.updateHealth(thc.SimpleCopy(), thc.Target, false, false)
+			hc.updateHealth(thc, thc.SimpleCopy(), thc.Target, false, false)
 		}
 
 		// Streaming RPC failed e.g. because vttablet was restarted or took too long.

--- a/go/vt/discovery/topology_watcher_test.go
+++ b/go/vt/discovery/topology_watcher_test.go
@@ -33,6 +33,7 @@ import (
 	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
 	"vitess.io/vitess/go/vt/topo"
 	"vitess.io/vitess/go/vt/topo/memorytopo"
+	"vitess.io/vitess/go/vt/topo/topoproto"
 )
 
 func checkOpCounts(t *testing.T, prevCounts, deltas map[string]int64) map[string]int64 {
@@ -834,6 +835,15 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 			return nil
 		})
 		require.NoError(t, err)
-		hc.updateHealth(th, prevTarget, false, false)
+		// Refetch the registered tabletHealthCheck each iteration: the topology
+		// watcher concurrently replaces it, and during the remove-to-add gap of
+		// ReplaceTablet there is none registered at all.
+		hc.mu.Lock()
+		thc := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(tablet1.Alias))]
+		hc.mu.Unlock()
+		if thc == nil {
+			continue
+		}
+		hc.updateHealth(thc, th, prevTarget, false, false)
 	}
 }

--- a/go/vt/discovery/topology_watcher_test.go
+++ b/go/vt/discovery/topology_watcher_test.go
@@ -806,16 +806,8 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 	hc.topoWatchers[0].loadTablets()
 	require.NoError(t, err)
 
-	// We want to run updateHealth with arguments that always
-	// make it trigger load Tablets.
-	th := &TabletHealth{
-		Tablet: tablet1,
-		Target: &querypb.Target{
-			Keyspace:   "keyspace",
-			Shard:      "shard",
-			TabletType: topodatapb.TabletType_REPLICA,
-		},
-	}
+	// We want to run updateHealth with a previous target that always
+	// makes it trigger load Tablets.
 	prevTarget := &querypb.Target{
 		Keyspace:   "keyspace",
 		Shard:      "shard",
@@ -844,6 +836,6 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 		if thc == nil {
 			continue
 		}
-		hc.updateHealth(thc, th, prevTarget, false, false)
+		hc.updateHealth(thc, prevTarget, false, false)
 	}
 }

--- a/go/vt/discovery/topology_watcher_test.go
+++ b/go/vt/discovery/topology_watcher_test.go
@@ -33,7 +33,6 @@ import (
 	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
 	"vitess.io/vitess/go/vt/topo"
 	"vitess.io/vitess/go/vt/topo/memorytopo"
-	"vitess.io/vitess/go/vt/topo/topoproto"
 )
 
 func checkOpCounts(t *testing.T, prevCounts, deltas map[string]int64) map[string]int64 {
@@ -830,9 +829,7 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 		// Refetch the registered tabletHealthCheck each iteration: the topology
 		// watcher concurrently replaces it, and during the remove-to-add gap of
 		// ReplaceTablet there is none registered at all.
-		hc.mu.Lock()
-		thc := hc.healthByAlias[tabletAliasString(topoproto.TabletAliasString(tablet1.Alias))]
-		hc.mu.Unlock()
+		thc := hc.registeredHealthCheck(tablet1.Alias)
 		if thc == nil {
 			continue
 		}

--- a/go/vt/discovery/topology_watcher_test.go
+++ b/go/vt/discovery/topology_watcher_test.go
@@ -818,6 +818,7 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 	// the tablet in the health check, but health check has the mutex acquired
 	// already because it is calling updateHealth.
 	// updateHealth itself will be stuck trying to send on the shared channel.
+	updates := 0
 	for i := range 10 {
 		// Update the port of the tablet so that when update Health asks topo watcher to
 		// refresh the tablets, it finds an update and tries to replace it.
@@ -834,5 +835,7 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 			continue
 		}
 		hc.updateHealth(thc, prevTarget, false, false)
+		updates++
 	}
+	require.NotZero(t, updates, "expected at least one updateHealth call")
 }

--- a/go/vt/discovery/topology_watcher_test.go
+++ b/go/vt/discovery/topology_watcher_test.go
@@ -818,7 +818,6 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 	// the tablet in the health check, but health check has the mutex acquired
 	// already because it is calling updateHealth.
 	// updateHealth itself will be stuck trying to send on the shared channel.
-	updates := 0
 	for i := range 10 {
 		// Update the port of the tablet so that when update Health asks topo watcher to
 		// refresh the tablets, it finds an update and tries to replace it.
@@ -835,7 +834,5 @@ func TestDeadlockBetweenTopologyWatcherAndHealthCheck(t *testing.T) {
 			continue
 		}
 		hc.updateHealth(thc, prevTarget, false, false)
-		updates++
 	}
-	require.NotZero(t, updates, "expected at least one updateHealth call")
 }

